Frameworks/Foundation/NSURLConnection.mm

```diff
@@ -19,168 +20,289 @@
 */
-- (id)initWithRequest:(id)request delegate:(id)delegate startImmediately:(BOOL)startLoading {
-    if (self = [super init]) {
-        @try {
-            [self _setRequest:request];
-        } @catch (NSException* exception) {
-            [self release];
-            return nil;
-        }
+- (instancetype)initWithRequest:(NSURLRequest*)request delegate:(id)delegate {
+    return [self initWithRequest:request delegate:delegate startImmediately:YES];
+}
 
-        _delegate = [delegate retain];
+/**
+ @Status Interoperable
+*/
+- (instancetype)initWithRequest:(NSURLRequest*)request delegate:(id)delegate startImmediately:(BOOL)startImmediately {
+    if (self = [super init]) {
+        _originalRequest = [request copy];
+        _delegate = delegate;
 
-        if (startLoading) {
+        if (startImmediately) {
             [self start];
         }
     }
     return self;
 }
 
-- (BOOL)_setRequest:(NSURLRequest*)request {
-    NSURLRequest* copiedRequest = [[request copy] autorelease];
-
-    id cls = [NSURLProtocol _URLProtocolClassForRequest:copiedRequest];
-    if (!cls || ![cls canInitWithRequest:copiedRequest]) {
-        return NO;
-    }
-
-    _protocol = [[cls alloc] initWithRequest:copiedRequest
-                              cachedResponse:[[NSURLCache sharedURLCache] cachedResponseForRequest:copiedRequest]
-                                      client:self];
-    if (!_protocol) {
-        return NO;
-    }
-
-    _request = [copiedRequest retain];
-
-    return YES;
-}
-
 /**
  @Status Interoperable
 */
-- (id)initWithRequest:(id)request delegate:(id)delegate {
-    return [self initWithRequest:request delegate:delegate startImmediately:YES];
++ (void)sendAsynchronousRequest:(NSURLRequest*)request
+                          queue:(NSOperationQueue*)queue
+              completionHandler:(void (^)(NSURLResponse*, NSData*, NSError*))handler {
+    if (request) {
+        NSURLConnection* connection = [[[self alloc] initWithRequest:request
+                                                            delegate:[_NSURLConnection_DataDelegate delegateWithQueue:queue handler:handler]
+                                                    startImmediately:NO] autorelease];
+        [connection setDelegateQueue:queue];
+        [connection start];
+    }
 }
 
 /**
  @Status Interoperable
 */
 - (void)dealloc {
-    [_request release];
-    [_protocol release];
-    [_delegate release];
-    [_response release];
+    [_originalRequest release];
     [super dealloc];
 }
 
 /**
  @Status Interoperable
 */
 - (void)start {
-    if (!_didRetain) {
-        [self retain];
-        _didRetain = TRUE;
+    @synchronized(self) {
+        NSURLProtocol* protocol = _protocolForRequest(_originalRequest, self);
+        if (!protocol) {
+            TraceError(TAG, L"[NSURLConnection start] could not create a NSURLProtocol for its NSURLRequest.");
+            return;
+        }
+        _protocol = protocol;
+
+        if (_delegateQueue.get()) {
```

Comment:
nit: prefer `if(_delegateQueue)`; StrongId/AutoId have an `operator bool` that will not require the pointer to be returned to the caller.